README
Authors: Don Conrad, Avinash Ramu ( aramu at genetics dot wustl dot edu ) and Reed A. Cartwright.

## RELEASE NOTES
v0.5.4
added GPL v3
updated output fields for indels, snps to be the same 

v0.5.3
removed 'X' allele in VCF op. VCF can be indexed by Tabix, IGVTools and used in Annovar.
added region based denovo calling on BCF files, invoked with --region flag
added vcf parser for denovo calling, invoked with --vcf flag


v0.5.2
Added read-depth, posterior-probability filters.
Output number of sites in the BCF and number of sites passing filters.
Modified paired caller output.

v0.5.1
Fixed bug in triallelic configuration. Some trialleic denovo configurations were being called incorrectly.

## COMPILING 
Compilation of DeNovoGear requires CMake.  You can download CMake installers from the CMake
website <http://www.cmake.org/cmake/resources/software.html>.  Most Linux
distributions allow you to install CMake using their package software.

Compiling and Installing on Linux:
        `tar xvzf denovogear*.tar.gz`
        `cd denovogear*/build`
        `cmake ..`
        `make`
        `sudo make install`

Creating Packages:
        `make package`
        `make package_source`

## BINARIES
Binaries are available for MacOSX and for Linux.

## RUNNING THE CODE

### Finding Denovo Mutations in trios and pairs.

DeNovoGear takes in a PED file and a BCF file as input. The PED file describes the relationship between the samples and the BCF file contains the sequencing information for every locus.

#### usage:
	     `./denovogear auto dnm --ped sample.ped --bcf sample.bcf`

#####  about sample.bcf:
BCF files can be generated from the alignment using the samtools mpileup 
command. The command to generate a bcf file from sample.bam is:
	`samtools mpileup -gDf reference.fa sample.bam > sample.bcf`

The -D option of the samtools mpileup command retains the per-sample read depth 
which is preferred by denovogear as it helps to filter out sites without a minimum number of reads(but note that DNG will work without per-sample RD information, in which case the RD tag encodes the average read depth information). The -g option computes genotype likelihoods and produces a compressed bcf output and the -f option is used to indicate the reference fasta file against which the alignment was built. A sample BCF file 'sample_CEU.bcf' is included in the distribution.

#####  about sample.ped: 
The PED file contains information about the trios present in the BCF file. 
Please make sure that all the members of the trios specified in the PED file 
are present in the BCF file. The PED file can be used to specify a subset of 
individuals for analysis from the BCF (in other words not every sample in the BCF needs to 
be represented in the PED file).

The PED file is a tab delimited file. The first six columns of the PED file are 
mandatory, these are Family ID, Individual ID, Paternal ID, Maternal ID, 
Sex (1 = male; 2 = female; other = unknown) and Phenotype. Denovogear makes use of the first four columns. The sample ID's in the PED file need to be exactly as they appear in the BCF file header. Sample order within the PED file does not matter, as family relationships are completely specified by the value of the child/mother/father fields in each row.
 
For example, a single line in the PED file that specifies a trio looks like:

CEU	NA12878_vald-sorted.bam.bam	NA12891_vald-sorted.bam.bam	NA12892_vald-sorted.bam.bam	2	2

An example PED file, sample_CEU.ped, is included in the distribution directory. 

##### about "snp_lookup.txt" and "indel_lookup.txt": 
These are tables with precomputed priors (and other useful numbers) for all possible 
trio configurations, under the null (no mutation present) and alternative (true de novo). 
The default tables are generated during each program run using a prior of 
1 x 10 ^-8 /bp/generation on the haploid germline point mutation rate, 
and 1 x 10 ^-9 /bp/generation on the haploid germline indel mutation rate. 

If you wish to change the default point or indel mutation rates use the --snp_mrate 
or --indel_mrate switches respectively. 

For example
	     `./denovogear dnm auto --ped sample.ped --bcf sample.bcf --snp_mrate 2e-10 --indel_mrate 1e-11`
	            [OR]
	     `./denovogear dnm auto --ped sample.ped --vcf sample.vcf --snp_mrate 2e-10 --indel_mrate 1e-11`		    

The indel mutation rate prior is calculated based on the length of the insertion or deletion event, 
separate models are used for insertions and deletions. The two models are based on the indel observations from the 1000Genomes phase 1 data.

The insertion mutation rate is modeled using the function
 	log (mrate) = mu_scale * (-22.8689 - (0.2994 * insertionLength))

The deletion mutation rate is modeled using the function
	log (mrate) = mu_scale * (-21.9313 - (0.2856 * deletionLength))

Note that a constant factor is used to scale the mutation rate, it is set to 1.0 
by default and can be set using the switch --mu_scale. This provides the users to scale the mutation rate prior according to their data-set.

For example, 
	     `./denovogear dnm auto --ped sample.ped --bcf sample.bcf --mu_scale 3`
	     		   [OR]
   	     `./denovogear dnm auto --ped sample.ped --vcf sample.vcf --mu_scale 3`	


#### OUTPUT FORMAT for TRIOS

The output format is a single row for each putative de novo mutation (DNM), with the following fields

        1. Event type (POINT MUTATION or INDEL).
        2. CHILD_ID - sample ID of trio-offspring with the DNM.
        3. chr - chromosome. 
        4. pos - physical Position. 
        5. ref - base present in reference sequence at this position (from BCF).
        6. alt - comma separated list of alternate non-reference alleles called on at-least one sample (from BCF).
        7. maxlike_null  - likelihood of the most likely mendelian-compatible genotype configuration.
        8. pp_null - posterior probability of the most likely mendelian-compatible genotype configuration.        
        9. tgt_null(child/mom/dad)  - genotypes of the most likely mendelian-compatible configuration.
        10. snpcode - code that indicates whether the configuration shown in field 6 is monomorphic (1) or contains variation (2)(internal filters, can be ignored).
        11. code - This field seems to be redundant to field 7, except the codes are (6) and (9).(internal filters, can be ignored).
        12. maxlike_dnm - likelihood of the most likely DENOVO genotype configuration.
        13. pp_dnm - posterior probability of the most likely DENOVO genotype configuration. Range = [0-1], a value closer to 1 indicates higher probability of observing a denovo event at this position. This is the field that is used to rank the calls.
        14. tgt_dnm(child/mom/dad)  - genotypes of the most likely mendelian-compatible configuration.
        15. lookup - Code that indicates if the most likely DNM is a transition (4) or transversion (5) (for development use only).
        16. flag - Flag that indicates whether the data for the site passed internal QC thresholds (for development use only).
        17-19. Read depth of child, parent 1 and parent 2. 
        20-22. Root mean square of the mapping qualities of all reads mapping to the site for child, parent 1 and parent 2. Currently these values are the same for all samples when using BCF as the input format.

Fields 17-22 are meant for filtering out low quality sites. 

### Separate models for the X chromosome

Denovogear has separate models for autosomes, X chromosome in male offspring and X chromosome in female offspring. To use this model, create separate BCFs for the X chromosome.

#### Autosomes model usage

        `./denovogear dnm auto --ped sample.ped --bcf sample.bcf`
		      [OR]
        `./denovogear dnm auto --ped sample.ped --vcf sample.vcf`

#### X in male offspring model usage 

        `./denovogear dnm XS --ped sample.ped --bcf sample.bcf --region X`
		      [OR]
        `./denovogear dnm XS --ped sample.ped --vcf sample.X.vcf` (only BCF allows for random access)
#### X in female offspring model usage 

        `./denovogear dnm XD --ped sample.ped --bcf sample.bcf --region X`
		      [OR]
        `./denovogear dnm XD --ped sample.ped --vcf sample.X.vcf`		   

### PAIRED SAMPLE ANALYSIS 
DNG can be used to analyze paired samples, it is run the same way as for trios the only difference being the way samples are specified in the PED file,

#### Usage:
 
        `./denovogear dnm auto --ped paired.ped --bcf sample.bcf`
		      [OR]
        `./denovogear dnm auto --ped paired.ped --vcf sample.vcf`

About the arguments, 
	
	1. paired.ped is a ped file containing the family-name and the name of the two samples in the bcf file. The last three columns are mandated by the 	PED format but are ignored by the program. An example line looks like
		YRI    NA19240_blood_vald-sorted.bam.bam NA19240_vald-sorted.bam.bam   0       0       0
	2. sample.bcf[vcf] is a BCF[VCF] file containing both the samples. 

A sample PED file sample_paired.ped for paired sample analysis is provided with the package.

#### OUTPUT FORMAT for Paired Sample Analysis
The output format is a single row for each putative paired denovo mutation(DNM), with the following fields

        1. Event type (POINT MUTATION or INDEL).
        2. TUMOR_ID - sample ID of the 'TUMOR' sample.
        3. NORMAL_ID - sample ID of the 'NORMAL' sample.
        4. chr - chromosome. 
        5. pos - physical Position. 
        6. ref - base present in reference sequence at this position (from BCF).
        7. alt - comma separated list of alternate non-reference alleles called on at-least one sample (from BCF).
        8. maxlike_null  - likelihood of the most likely compatible genotype configuration.
        9. pp_null - posterior probability of the most likely compatible genotype configuration.        
        10. tgt_null(normal/tumor)  - genotypes of the most likely compatible configuration.
        11. maxlike_dnm  - likelihood of the most likely denovo genotype configuration.
        12. pp_dnm - posterior probability of the most likely denovo genotype configuration, a value closer to 1 indicates strong evidence of a denovo event. This is the field that is used to rank the calls.        
        13. tgt_dnm(normal/tumor)  - genotypes of the most likely denovo configuration.
        14-15. Read depth of tumor, normal samples. 
        16-17. Root mean square of the mapping qualities of all reads mapping to the site for tumor, normal samples.
        18-19. null_snpcode, dnm_snpcode - snpcode is a field used to classify the genotype configurations for the null and alternate case. 1 stands for hom/hom, 2 stands for het/hom, 3 stands for hom/het and 4 stands for het/het.


### PHASER 
DNG can be used to obtain parental phasing information for Denovo Mutations where phase informative sites are present. This is done by looking at reads which cover both the denovo base and a phase informative positions. Phase informative positions lie within a certain window from the denovo site, the default value is 1000 but it can be set by the user. The phaser is under development, please do let us know if you find any bugs.

#### Usage:

        `./denovogear phaser --dnm dnm_f --pgt gts_file --bam alignment --window 1000`

About the arguments, 

	1. 	"dnm" is the list of DNMs whose parental origin is to be determined. It is a tab delimited file of the format
	  	chr pos inherited_base mutant_base
	2.	"pgt" contains the genotypes of the child and the parents at SNP sites. It is a tab delimited file of the format
		chr pos child_GT parent1_GT parent2_GT
	3. 	"bam" is the alignment file (.bam) of the child. 
	4. 	"window" is an optional argument which is the maximum distance between the DNM and a phasing site. The default value is 1000. 

#### Output
	DNM_pos 1:182974758     INHERITED G     VARIANT A
        HAP POS 182974328 p1: CC p2: TC Number of denovo-phasing pairs found: 0
        HAP POS 182974572 p1: CC p2: TC Number of denovo-phasing pairs found: 2
                Base at DNM position: A Base at phasing position: C      INFERRED PARENT OF ORIGIN for DNM: p1 SUPPORTING READ COUNT: 3
                Base at DNM position: G Base at phasing position: T      INFERRED PARENT OF ORIGIN for DNM: p1 SUPPORTING READ COUNT: 3
        HAP POS 182974598 p1: GG p2: CG Number of denovo-phasing pairs found: 2
                Base at DNM position: A Base at phasing position: G      INFERRED PARENT OF ORIGIN for DNM: p1 SUPPORTING READ COUNT: 1
                Base at DNM position: G Base at phasing position: C      INFERRED PARENT OF ORIGIN for DNM: p1 SUPPORTING READ COUNT: 2
        HAP POS 182974602 p1: TT p2: CT Number of denovo-phasing pairs found: 2
                Base at DNM position: A Base at phasing position: T      INFERRED PARENT OF ORIGIN for DNM: p1 SUPPORTING READ COUNT: 1
                Base at DNM position: G Base at phasing position: C      INFERRED PARENT OF ORIGIN for DNM: p1 SUPPORTING READ COUNT: 1
        HAP POS 182974707 p1: TT p2: CT Number of denovo-phasing pairs found: 2
                Base at DNM position: A Base at phasing position: T      INFERRED PARENT OF ORIGIN for DNM: p1 SUPPORTING READ COUNT: 25
                Base at DNM position: G Base at phasing position: C      INFERRED PARENT OF ORIGIN for DNM: p1 SUPPORTING READ COUNT: 24
        HAP POS 182974750 p1: AA p2: TA Number of denovo-phasing pairs found: 2
                Base at DNM position: A Base at phasing position: A      INFERRED PARENT OF ORIGIN for DNM: p1 SUPPORTING READ COUNT: 45
                Base at DNM position: G Base at phasing position: T      INFERRED PARENT OF ORIGIN for DNM: p1 SUPPORTING READ COUNT: 47


	For each DNM, a list of of eligible phasing sites from the genotypes file is obtained, these are the loci which lie within the phasing window and which do not have a het/het GT configuration for the parents and where the child is het. The number of DNM and phasing  allele combinations seen in the read level data is output by the program. For the phasing sites, the inferred parental origin is displayed.

For example if the base at the phasing site is T and the parental genotypes are P1:CC and p2:TC at this site, then the parent of origin of this base is p2. By looking at the base in the denovo position on this read it is possible to infer the parent of origin of the denovo mutation. 

A sample list of dnms file, sample_phasing_dnm_f is provided. Also a sample genotypes file sample_phasing_GTs_f is provided for reference.

Please feel free to contact the authors about any concerns/comments.	

###General options for TRIOs and Paired Sample Calling
--snp_mrate:     Mutation rate prior for SNPs. [1e-8]
--indel_mrate:   Mutation rate prior for INDELs. [1e-9]
--pair_mrate:    Mutation rate prior for paired sample analysis. [1e-9]
--indel_mu_scale:        Scaling factor for indel mutation rate. [1]
--pp_cutoff:     Posterior probability threshold. [0.0001]
--rd_cutoff:     Read depth filter, sites where either one of the sample have read depth less than this threshold are filtered out. [10
--region: region of the BCF file over which to perform denovo calling. [string of the form "chr:start-end"]

## ACKNOWLEDGEMENTS
DeNovoGear uses the Samtools libraries to parse BCF files, we thank the Heng Li and the rest of the authors of Samtools for making this resource available to developers.


CHANGELOG.md
## Release Notes

### develop

* Lots of new features and bug fixes are added to the develop branch, which is generally stable. Consider trying it out.

#### latest changes in develop

* CHANGE: HTSLIB 1.3.1+ is now required
* FEATURE: added `dng treecall` v2
* FEATURE: new k-alleles model for `dng call` that improves mutation calling
* BUGFIX: memory leak fixed with when using BCFtools parsing functions
* FEATURE: improved `dng pileup` now supports slurping regions and filter in number of alleles
* FEATURE: added region specification with BED files when using VCF on `dng call` 
* FEATURE: added support for AD/TAD format in `dng call` and `dng loglike`
* FEATURE: updated version of external dependencies
* FEATURE: improved support for regions in BAM files
* FEATURE: `dng dnm` can now handle indels 
* BUGFIX: fixed segfault found in some cases when using `GCC 4.8` 
* FEATURE: improved compiling on Apple
* FEATURE: optimized peeler implementation and interface

### v1.1.1

* FEATURE: When building with cmake, users have the option to download and install dependencies.
Just run `cmake -DBUILD_EXTERNAL_PROJECTS=1 ..`
* FEATURE: `dng help` now fully implemented
* FEATURE: more information added to `dng call` output
* BUGFIX: When running `dng call` with HTSLIB 1.2.1, the following error message was being
emitted: "FIXME: dirty header not synced".  This has been fixed.
* BUGFIX: `dng call` now outputs correct 1-based site locations.
* BUGFIX: a segfault was fixed in `dng dnm` and `dng phaser` caused by invalid commandline arguments
* CHANGE: HTSLIB 1.2+ is now required.
* Miscellaneous improvements to the build system

### v1.1

* Main program now called 'dng'
* Added experimental 'dng call' module.
* DeNovoGear now requires HTSLIB 1+, CMake 3.1+, Boost 1.47+, and Eigen 3+.

### v1.0

* made changes to indel_mrate parameter
* better indenting
* mu_scale scales indel mutation rate linearly

### v0.5.4

* added GPL v3
* updated output fields for indels, snps to be the same

### v0.5.3

* removed 'X' allele in VCF op. VCF can be indexed by Tabix, IGVTools and used in Annovar.
* added region based denovo calling on BCF files, invoked with --region flag
* added vcf parser for denovo calling, invoked with --vcf flag

### v0.5.2

* Added read-depth, posterior-probability filters.
* Output number of sites in the BCF and number of sites passing filters.
* Modified paired caller output.

### v0.5.1

* Fixed bug in triallelic configuration.
* Some trialleic denovo configurations were being called incorrectly.

### v0.5

* switched to cmake
* separate models for X chromosome calling in male offspring, female offspring

### v0.4

* paired sample analysis
* parental phaser

### v0.3

* INDEL length based mutation prior
* look at additional trio configs compared to initial version

### v0.2

* incorporated BCF parser
* denovo INDEL calling


src/dng-dnm.h
#ifndef DNG_DNM_H_
#define DNG_DNM_H_

#include <iostream>
#include <string>
#include <vector>
#include <stdlib.h>
#include <getopt.h>
#include <memory> 
#include <assert.h> 

#include <dng/hts/bcf.h>

#include "version.h"
#include "lookup.h"
#include "makeLookup.h"


/*
int RD_cutoff = 10; // cutoff for the read depth filter
double PP_cutoff = 0.0001; // posterior probability cutoff

double indel_mrate = 0; // indel mutation prior is based on a log linear model unless specified by the user.
double snp_mrate = 1e-8; // snp mutation prior
double poly_rate = 1e-3; // polymorphism rate - used in prior calculations
double pair_mrate = 1e-9; // mutation prior for paired samples
double mu_scale = 1.0; // scaling factor for indel priors
*/


//int callDenovoFromBCF(std::string ped_file, std::string bcf_file, std::string op_vcf_f, 
//		      std::string model, bool is_vcf, std::string region);

/*
void usage() {
  std::cerr << "\nUsage:\n";
    std::cerr << "Autosomes:\n";
    std::cerr << "\tdng dnm auto --bcf bcf_f --ped ped_f [OR] dng dnm auto --vcf vcf_f --ped ped_f\n";
    std::cerr << "X chromosome in male offspring:\n";
    std::cerr << "\tdng dnm XS --bcf bcf_f --ped ped_f [OR] dng dnm XS --vcf vcf_f --ped ped_f\n";
    std::cerr << "X chromosome in female offspring:\n";
    std::cerr << "\tdng dnm XD --bcf bcf_f --ped ped_f [OR] dng dnm XD --vcf vcf_f --ped ped_f\n";
    //cerr<<"Phaser:\n";
    //cerr<<"\tdng phaser --dnm dnm_f --pgt pgt_f --bam bam_f --window INT[1000]\n";
    std::cerr << "\nInput:\n";
    std::cerr << "DNM:\n";
    std::cerr << "--ped:\t Ped file to describe relationship between the samples.\n";
    std::cerr << "--bcf:\t BCF file, contains per-sample read depths and genotype likelihoods.\n";
    std::cerr << "--vcf:\t VCF file, contains per-sample read depths and genotype likelihoods.\n";
    std::cerr << "Phaser:\n";
    std::cerr << "--dnm: Tab delimited list of denovo mutations to be phased, format: chr pos inherited_base denovo_base.[example: 1 2000 A C]\n";
    std::cerr << "--pgt: Tab delimited genotypes of child and parents at SNP sites near denovo sites, format: chr pos GT_child GT_parent1 GT_parent2.[example: 1 2000 AC AC AA]\n";
    std::cerr << "--bam: alignment file (.bam) of the child.\n";
    std::cerr << "--window: optional argument which is the maximum distance between the DNM and a phasing site. The default value is 1000.\n";
    std::cerr << "\nOutput:\n";
    std::cerr << "--output_vcf:\t vcf file to write the output to.\n";
    std::cerr << "\nParameters:\n";
    std::cerr << "--snp_mrate:\t Mutation rate prior for SNPs. [1e-8]\n";
    std::cerr << "--indel_mrate:\t Mutation rate prior for INDELs. [1e-9]\n";
    std::cerr << "--pair_mrate:\t Mutation rate prior for paired sample analysis. [1e-9]\n";
    std::cerr << "--indel_mu_scale:\t Scaling factor for indel mutation rate. [1]\n";
    std::cerr << "--pp_cutoff:\t Posterior probability threshold. [0.0001]\n";
    std::cerr << "--rd_cutoff:\t Read depth filter, sites where either one of the sample have read depth less than this threshold are filtered out. [10]\n";
    std::cerr << "--region:\t Region of the BCF file to perform denovo calling. [string of the form \"chr:start-end\"]\n";
    std::cerr << std::endl;
    exit(1);
}
*/

/*
int mainDNG(int argc, char *argv[]) {
  std::string model;
  if(argc > 1) {
    model = argv[1];
  } else {
    usage();
  }
  std::string ped_f = "EMPTY"; // ped file
  std::string bcf_f = "EMPTY"; // bcf/vcf file
  std::string op_vcf_f = "EMPTY"; // vcf output -- optional
  bool is_vcf = false;
  std::string region = "NULL";
  // Read in Command Line arguments
    while(1) {
        int option_index = 0;
        static struct option long_options[] = {
            {"ped", 1, 0, 0},
            {"bcf", 1, 0, 1},
            {"snp_mrate", 1, 0, 2},
            {"indel_mrate", 1, 0, 3},
            {"poly_rate", 1, 0, 4},
            {"pair_mrate", 1, 0, 5},
            {"indel_mu_scale", 1, 0, 6},
            {"output_vcf", 1, 0, 7},
            {"pp_cutoff", 1, 0, 8},
            {"rd_cutoff", 1, 0, 9},
            {"h", 1, 0, 10},
            {"vcf", 1, 0, 11},
            {"region", 1, 0, 12},
            {0, 0, 0, 0}
        };
        int c = getopt_long(argc - 1, argv + 1, "", long_options, &option_index);
        if(c == -1) {
            break;
        }
        switch(c) {
        case 0:
            ped_f = optarg;
            break;
        case 1:
            bcf_f = optarg;
            break;
        case 2:
            snp_mrate = atof(optarg);
	    std::cerr << "\nSNP mutation rate: " << snp_mrate;
            break;
        case 3:
            indel_mrate = atof(optarg);
	    std::cerr << "\nindel mutation rate: " << indel_mrate;
            break;
        case 4:
            poly_rate = atof(optarg);
	    std::cerr << "\npolymorphism rate: " << poly_rate;
            break;
        case 5:
            pair_mrate = atof(optarg);
	    std::cerr << "\npaired-sample mutation rate: " << pair_mrate;
            break;
        case 6:
            mu_scale = atof(optarg);
	    std::cerr << "\nindel mutation rate scaling factor: " << mu_scale;
            break;
        case 7:
            op_vcf_f = optarg;
	    std::cerr << "\noutput vcf file: " << op_vcf_f;
            break;
        case 8:
            PP_cutoff = atof(optarg);
	    std::cerr << "\nposterior probability cutoff: " << PP_cutoff;
            break;
        case 9:
            RD_cutoff = atoi(optarg);
	    std::cerr << "\nread depth filter: " << RD_cutoff;
            break;
        case 10:
            usage();
            break;
        case 11:
            bcf_f = optarg;
            is_vcf = true;
            break;
        case 12:
            region = optarg;
            break;
        default:
            usage();
        }
    }

    if((bcf_f == "EMPTY") || (ped_f == "EMPTY")) {
      std::cerr << "ERROR ! Please specify both the PED file and BCF file !"
             "Exiting!\n";
      //usage();
    }

    // Create lookup table and read ped, BCF files. Separate for autosomes, X in males and X in females.
    if(model ==  "auto" || model == "XS" || model == "XD") {
        callDenovoFromBCF(ped_f, bcf_f, op_vcf_f, model, is_vcf, region);
    } else {
        usage();
    }

    std::cerr << "\nDone !" << std::endl;
    exit(0);
}
*/

/*
int callMakeSNPLookup(lookup_table_t &tgtSNP, lookup_snp_t &lookupSNP, std::string model) {
  std::vector<string> tmp;
    for(int l = 0; l < 10; l++) {
        tmp.push_back("NA");
    }
    for(int l = 0; l < 100; l++) {
        tgtSNP.push_back(tmp);
    }


    if(model == "auto") { makeSNPLookup(snp_mrate, poly_rate, tgtSNP, lookupSNP); }
    else if(model == "XS") { makeXSSNPLookup(snp_mrate, poly_rate, tgtSNP, lookupSNP); }
    else if(model == "XD") { makeXDSNPLookup(snp_mrate, poly_rate, tgtSNP, lookupSNP); }
    else {
        cerr << endl << "Invalid model for SNP lookup, exiting.";
        exit(1);
    }

    cerr << "\nCreated SNP lookup table\n";
    cerr << " First mrate: " << lookupSNP.mrate(1, 1) << " last: " << lookupSNP.mrate(100, 10) << endl;
    cerr << " First code: " << lookupSNP.code(1, 1) << " last: " << lookupSNP.code(100, 10) << endl;
    cerr << " First target string: " << tgtSNP[0][0] << " last: " << tgtSNP[99][9] << endl;
    cerr << " First tref: " << lookupSNP.tref(1, 1) << " last: " << lookupSNP.tref(100, 10) << endl;
    return 0;
}

int callMakeINDELLookup(lookup_table_t &tgtIndel, lookup_indel_t &lookupIndel, std::string model) {
  std::vector<std::string> tmp;
    for(int l = 0; l < 3; l++) {
        tmp.push_back("NA");
    }
    for(int l = 0; l < 9; l++) {
        tgtIndel.push_back(tmp);
    }


    if(model == "auto") { makeIndelLookup(poly_rate, tgtIndel, lookupIndel); }
    else if(model == "XS") { makeXSIndelLookup(poly_rate, tgtIndel, lookupIndel); }
    else if(model == "XD") { makeXDIndelLookup(poly_rate, tgtIndel, lookupIndel); }
    else {
      std::cerr << std::endl << "Invalid model for INDEL lookup, exiting.";
      exit(1);
    }

    std::cerr << "\nCreated indel lookup table";
    std::cerr << " First code: " << lookupIndel.code(1, 1) << " last: " << lookupIndel.code(9, 3) << std::endl;
    std::cerr << " First target string: " << tgtIndel[0][0] << " last: " << tgtIndel[8][2] << std::endl;
    std::cerr << " First prior: " << lookupIndel.priors(1, 1) << " last: " << lookupIndel.priors(9, 3) << std::endl;
    return 0;
}

int callMakePairedLookup(lookup_table_t &tgtPair, lookup_pair_t &lookupPair) {
  std::vector<std::string> tmp;
    for(int l = 0; l < 10; l++) {
        tmp.push_back("NA");
    }
    for(int l = 0; l < 10; l++) {
        tgtPair.push_back(tmp);
    }
    makePairedLookup(pair_mrate, tgtPair, lookupPair);
    std::cerr << "\nCreated paired lookup table" << std::endl;
    std::cerr << " First target string: " << tgtPair[0][0] << " last: " << tgtPair[9][9] << std::endl;
    std::cerr << " First prior " << lookupPair.priors(1, 1) << " last: " << lookupPair.priors(10, 10) << std::endl;
    return 0;
}

void writeVCFHeader(hts::bcf::File &vcfout, std::string &bcf_file, std::string &ped_file, std::vector<std::string> &samples) {

  vcfout.AddHeaderMetadata("##fileformat=VCFv4.1");
  vcfout.AddHeaderMetadata("##source=", PACKAGE_STRING);
  vcfout.AddHeaderMetadata("##input_bcf=", bcf_file);
  vcfout.AddHeaderMetadata("##input_ped=", ped_file);
  vcfout.AddHeaderMetadata("##cutoff_read_depth=", RD_cutoff);
  vcfout.AddHeaderMetadata("##cutoff_posterior_probability=", PP_cutoff);
  vcfout.AddHeaderMetadata("##mutation_rate_snp=", snp_mrate);
  vcfout.AddHeaderMetadata("##mutation_rate_indel=", indel_mrate);
  vcfout.AddHeaderMetadata("##mutation_rate_paired=", pair_mrate);
  vcfout.AddHeaderMetadata("##mutation_rate_polymorphism=", poly_rate);
  vcfout.AddHeaderMetadata("##mutation_rate_indel_scaling_constant=", mu_scale);

#ifndef NEWVCFOUT
  // Old format, doesn't work well with htslib
  vcfout.AddHeaderMetadata("##INFO=<ID=RD_MOM,Number=1,Type=Integer,Description=\"Read depth for MOM\">");
  vcfout.AddHeaderMetadata("##INFO=<ID=RD_DAD,Number=1,Type=Integer,Description=\"Read depth for DAD\">");
  vcfout.AddHeaderMetadata("##INFO=<ID=RD_NORMAL,Number=1,Type=Integer,Description=\"Read depth for Normal sample in Paired Sample Analysis\">");
  vcfout.AddHeaderMetadata("##INFO=<ID=MQ_MOM,Number=1,Type=Integer,Description=\"Mapping quality for MOM\">");
  vcfout.AddHeaderMetadata("##INFO=<ID=MQ_DAD,Number=1,Type=Integer,Description=\"Mapping quality for DAD\">");
  vcfout.AddHeaderMetadata("##INFO=<ID=MQ_NORMAL,Number=1,Type=Integer,Description=\"Mapping quality for Normal sample in Paired Sample Analysis\">");
  vcfout.AddHeaderMetadata("##FORMAT=<ID=NULL_CONFIG(child/mom/dad),Number=1,Type=String,Description=\"NULL trio configuration\">");
  vcfout.AddHeaderMetadata("##FORMAT=<ID=PP_NULL,Number=1,Type=Float,Description=\"Posterior probability for the NULL configuration\">");
  vcfout.AddHeaderMetadata("##FORMAT=<ID=ML_NULL,Number=1,Type=Float,Description=\"Maximum Likelihood for the NULL configuration\">");
  vcfout.AddHeaderMetadata("##FORMAT=<ID=ML_DNM,Number=1,Type=Float,Description=\"Maximum Likelihood for the DNM configuration\">");
  vcfout.AddHeaderMetadata("##INFO=<ID=SNPcode,Number=1,Type=Integer,Description=\"Code that shows whether DNM configuration shown is monomorphic or contains variation\">");
  vcfout.AddHeaderMetadata("##INFO=<ID=INDELcode,Number=1,Type=Integer,Description=\" \">"); //// Needs Detail
  vcfout.AddHeaderMetadata("##INFO=<ID=PairSNPcode,Number=1,Type=Integer,Description=\" \">"); //// Needs Detail
  vcfout.AddHeaderMetadata("##INFO=<ID=code,Number=1,Type=Integer,Description=\" \">"); //// Needs Detail
  vcfout.AddHeaderMetadata("##FORMAT=<ID=DNM_CONFIG(child/mom/dad),Number=1,Type=String,Description=\"DNM trio configuration\">");
  vcfout.AddHeaderMetadata("##FORMAT=<ID=PP_DNM,Number=1,Type=Float,Description=\"Posterior probability for the DNM configuration\">");
  vcfout.AddHeaderMetadata("##FORMAT=<ID=RD,Number=1,Type=Integer,Description=\"Read Depth of the child\">");
  vcfout.AddHeaderMetadata("##FORMAT=<ID=MQ,Number=1,Type=Integer,Description=\"Mapping quality of the child\">");
  vcfout.AddHeaderMetadata("##FORMAT=<ID=RD_T,Number=1,Type=Integer,Description=\"Read Depth of the tumor sample\">");
  vcfout.AddHeaderMetadata("##FORMAT=<ID=MQ_T,Number=1,Type=Integer,Description=\"Mapping quality of the normal sample\">");
  vcfout.AddSample(samples[0].c_str());
#else
  // Newer header, produces more standarized VCF output
  vcfout.AddHeaderMetadata("##FORMAT=<ID=RD,Number=1,Type=Integer,Description=\"Read Depth\">");
  vcfout.AddHeaderMetadata("##FORMAT=<ID=MQ,Number=1,Type=Integer,Description=\"Mapping quality\">");
  for(std::string sample : samples)
    vcfout.AddSample(sample.c_str());
#endif
  vcfout.WriteHeader();
}
*/

#endif
